core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala

```diff
@@ -240,6 +239,40 @@ class BrokerMetadataListenerTest {
     }
   }
 
+  @Test
+  def testNotSnapshotAfterMetadataVersionChangeBeforePublishing(): Unit = {
+    val snapshotter = new MockMetadataSnapshotter()
+    val listener = newBrokerMetadataListener(snapshotter = Some(snapshotter),
+      maxBytesBetweenSnapshots = 1000L)
+
+    updateFeature(listener, feature = MetadataVersion.FEATURE_NAME, MetadataVersion.latest.featureLevel(), 100L)
+    listener.getImageRecords().get()
+    assertEquals(-1L, snapshotter.activeSnapshotOffset, "We won't generate snapshot before starting publishing")
+  }
+
+  @Test
+  def testSnapshotAfterMetadataVersionChangeWhenStarting(): Unit = {
```

Comment:
Amongst all these tests we want to ensure that a stray `handleSnaphot` event does not modify the logic we expected. Perhaps spy the MetadataListener and assert that no calls to `handleSnapshot` was made. This would ensure that the snaphot was triggered specifically because of version change and not because of some external event.

Reply:
In a real cluster context, `handleSnaphot` and `handleCommit` are invoked from the Raft listener, and in this test case, they can only be invoked by us. since we only invoke `handleCommit` in this test I don't think it's necessary to check `handleSnapshot`.